File: hw0.py
def palindrome(filename = "/usr/share/dict/words", letter = "all"):
   # fileNameSplit = filename.split('/')
    wordList = []
    inFile = open(filename, 'r')
    for line in inFile:
        alnumLine = (''.join(ch for ch in line if ch.isalnum())).lower()
        if letter == 'all' or alnumLine.startswith(letter):
            line = line.replace("\n", "")
            if alnumLine == alnumLine[:: -1]:
                wordList.append(line)
                print(line)
    return wordList    

def anagram(filename = "/usr/share/dict/words", target = False):
    inFile = open(filename, 'r')
    anaDict = {}
    if target == False:
        sortedAnaWord = ''
    else:
        sortedAnaWord = ''.join(sorted(target))
    for line in inFile:
        sortedLine = ''.join(sorted(line.lower()))
        sortedLine = ''.join(ch for ch in sortedLine if ch.isalnum());
        wordList = anaDict.get(sortedLine)
        if (sortedAnaWord == '') | (sortedLine == sortedAnaWord):
            if wordList is not None:
                wordList.append(line.strip())
                anaDict[sortedLine] = wordList;
            else :
                anaDict[sortedLine] = [line.strip()]
    maxCount = 0;
    for lineList in anaDict:
        if anaDict[lineList] is None:
            currentLength = 0;
        else:
            currentLength = len(anaDict[lineList])
        if maxCount < currentLength :
            maxCount = currentLength
            maxLine = [anaDict[lineList]]
        elif maxCount == currentLength :
            maxLine.append(anaDict[lineList])
    maxCount = 0;
    for line in maxLine[0]:
      print(line)

File: hw0.R
palindrome <- function(filename = "/usr/share/dict/words", letter = "all"){
  fileLines = readLines(filename)
  palindromeList = list()
  for(line in fileLines) {
    alnumLine = tolower(gsub("[^[:alnum:]]", "", line))
    if((letter == 'all' | strsplit(alnumLine, "")[[1]][1] == letter) & identical(alnumLine, paste(rev(strsplit(alnumLine, "")[[1]]), collapse = ""))){
      palindromeList = c(palindromeList, line)
    }
  }
  for(line in palindromeList){
    print(line)
  }
}

anagram <- function(filename = "/usr/share/dict/words", target = FALSE){
  fileLines = readLines(filename)
  anaList = list()
  maxVector = c()
  maxCount = 0
  if(target == FALSE){
    sortedAnaWord = ""
  }
  else {
    sortedAnaWord = paste(sort(strsplit(target, "")[[1]]), collapse = "")
  }
  for(line in fileLines) {
    alnumLine = strsplit(gsub("[^[:alnum:]]", "", line), "")[[1]]
    sortedAlnumLine = paste(sort(alnumLine), collapse = "")
    storedLine = anaList[[sortedAlnumLine]]
    alnumLine = paste(alnumLine, collapse = "")
    
    if((sortedAnaWord == "") | (sortedAlnumLine == sortedAnaWord)) {
      if(is.null(storedLine)) {
        anaList[[sortedAlnumLine]] = alnumLine
        count = 1
      }
      else {
        storedLine = c(storedLine, alnumLine)
        anaList[[sortedAlnumLine]] = storedLine
        count = length(storedLine)
      }
      if(maxCount < count) {
        maxCount = count
        maxVector = storedLine
      }
    }
  }
  print(maxVector)
}


